### ISSUE TITLE
Class injection on subclass causes build failure

### ISSUE THREAD

**Issue by rychere** [silverstripe/silverstripe-framework#8935]
## Versions affected

4.3.0

## Description

Replacing a subclass DataObject via injection will cause a build error.

The generation of a table for a DataObject occurs in the following steps in `SilverStripe\ORM\DatabaseAdmin->doBuild`:

1. All tables are identified by `ClassInfo::subclassesFor(DataObject::class);`
2. Code `singleton($dataClass)` generates a singleton based on each subclass. This will call the injector and return the injected class.
3. Replacement subclass table will be generated, original subclass table is not generated.

An error occurs in `SilverStripe\ORM\DatabaseAdmin->updateLegacyClassNames`
1. `updateLegacyClassNames` is called for all subclasses using `ClassInfo::subclassesFor(DataObject::class);`
2. `updateLegacyClassNames` calls `DataObject::get()`  with the subclass name, which is chained with `count()`. This executes a query on the original subclass table name.
3. Subclass table does not exist, causing the query to fail.

## Steps to Reproduce

1. Start a new project
2. Create a DataObject BaseClass with database fields
3. Create a DataObject ExtendedClass with database fields
4. Create a DataObject InjectedClass with database fields
5. Add an Injector.yml
```
---
Name: AnInjector
After:
  - '#requestprocessors'
  - '#coresecurity'
---

SilverStripe\Core\Injector\Injector:
  ExtendedClass:
    class: InjectedClass
```

Run a `dev/build`

*Result*
`Couldn't run query: SELECT DISTINCT count(DISTINCT "ExtendedClass"."ID") AS "Count" FROM "ExtendedClass" WHERE ("ExtendedClass"."ClassName" = ?)`

## Possible fix

Replacing 
`singleton($dataClass)`
in `SilverStripe\ORM\DatabaseAdmin->doBuild` with
`new $dataClass`
ensures a table is generated for the replaced subclass. I am not that familiar with the conventions / code in the core so am not sure whether this would be sufficient.

**Comment by michalkleiner** [comment]
I'd say `Injector` is the preferred way of getting an instance so it would rather be a fix at that end, not using `new`.

**Comment by ScopeyNZ** [comment]
- Does `InjectedClass` extend `ExtendedClass`?
- What's the `table_name` configuration for your `InjectedClass`?
- What's the rest of the error reported by your SQL engine?

It might be that the existing `table_name` is used because it's still checking configuration on `ExtendedClass` and not `InjectedClass`.

**Comment by rychere** [comment]
To be clear on what is going on:
 - the injector process in doBuild stops the table for `ExtendedClass` from being generated
 - the classname remapping process attempts to re-map the table for `ExtendedClass`

@ScopeyNZ 

- Occurs both when `InjectedClass` extends `ExtendedClass` and when it doesn't
- I omitted `table_name` and namespaces in the reproduceable steps for simplicity, but have seen no difference in a project that uses both namespaces and `table_name`.
- The table does not exist - `Uncaught SilverStripe\ORM\Connect\DatabaseException: Couldn't run query: SELECT DISTINCT count(DISTINCT "ExtendedClass"."ID") AS "Count" FROM "ExtendedClass" WHERE ("ExtendedClass"."ClassName" = ?) 42S02-1146: Table 'ss_test.extendedclass' doesn't exist`
- It is unclear whether the fields or relations on the classes have any impact

Basic example for reproduction:

```
diff --git a/app/_config/Injector.yml b/app/_config/Injector.yml
new file mode 100644
index 0000000..c3e8688
--- /dev/null
+++ b/app/_config/Injector.yml
@@ -0,0 +1,10 @@
+---
+Name: AnInjector
+After:
+  - '#requestprocessors'
+  - '#coresecurity'
+---
+
+SilverStripe\Core\Injector\Injector:
+  ExtendedClass:
+    class: InjectedClass
diff --git a/app/src/DataObjects/BaseClass.php b/app/src/DataObjects/BaseClass.php
new file mode 100644
index 0000000..cc5da7b
--- /dev/null
+++ b/app/src/DataObjects/BaseClass.php
@@ -0,0 +1,18 @@
+<?php
+
+use SilverStripe\ORM\DataObject;
+
+class BaseClass extends DataObject
+{
+
+    private static $db = [
+        'Title'    => 'Varchar(255)',
+        'Content'  => 'HTMLText',
+        'LinkHref' => 'Varchar(255)',
+        'LinkText' => 'Varchar(255)',
+               'SortOrder' => 'Int'
+    ];
+
+    private static $default_sort = 'SortOrder';
+
+}
diff --git a/app/src/DataObjects/ExtendedClass.php b/app/src/DataObjects/ExtendedClass.php
new file mode 100644
index 0000000..bad2b27
--- /dev/null
+++ b/app/src/DataObjects/ExtendedClass.php
@@ -0,0 +1,14 @@
+<?php
+
+use SilverStripe\ORM\DataObject;
+
+class ExtendedClass extends DataObject
+{
+
+    private static $db = [
+        'Field'    => 'Varchar(255)'
+    ];
+
+    private static $default_sort = 'SortOrder';
+
+}
diff --git a/app/src/DataObjects/InjectedClass.php b/app/src/DataObjects/InjectedClass.php
new file mode 100644
index 0000000..f032b70
--- /dev/null
+++ b/app/src/DataObjects/InjectedClass.php
@@ -0,0 +1,14 @@
+<?php
+
+use SilverStripe\ORM\DataObject;
+
+class InjectedClass extends DataObject
+{
+
+    private static $db = [
+        'FieldTwo'    => 'Varchar(255)'
+    ];
+
+    private static $default_sort = 'SortOrder';
+
+}
```

**Comment by ScopeyNZ** [comment]
Triaged as a bug - but it seems really low impact to me. Overriding a class like that instead of just using a `DataExtension` seems odd.

**Comment by tractorcow** [comment]
You'd need to build a brand new "replaces" config, so that the replacement class itself knew which class to replace.

Code which enumerated hierarchy (e.g. ClassInfo::ancestry()) would need to be rewritten to respect this, so that the parent/child pair (child replaces a parent in the hierarchy) only existed once (ideally, the replaced class).

Since the parent class is essentially ignored, you'd probably need to duplicate all configs (e.g. db / has_ones) in the subclass, since Config() uses uninherited search.

Long story short, supporting this is a huge amount of work and risk for little benefit. I would recommend extensions for any case this would otherwise be used.
